src/protocols/secure_channel/CASESession.cpp

```diff
@@ -125,14 +125,36 @@ static constexpr ExchangeContext::Timeout kSigma_Response_Timeout = System::Cloc
 
 CASESession::~CASESession()
 {
-    // Let's clear out any security state stored in the object, before destroying it.
-    Clear();
+    // This function zeroes out and resets the memory used by the object.
+    // It's done so that no security related information will be leaked.
+    mCommissioningHash.Clear();
+    PairingSession::Clear();
+    Crypto::ClearSecretData(mIPK);
+    AbortExchange();
 }
 
-void CASESession::Finish()
+bool CASESession::SanityCheck() const
 {
-    mCASESessionEstablished = true;
+    // This is called after handling a message, the session must be in a state of done or waiting for another message,
+    // otherwise the object may be leaked.
+    if (IsDone())
+        return true;
+
+    if (mRole == CryptoContext::SessionRole::kInitiator)
+    {
+        return mState == State::kSentSigma1 || mState == State::kSentSigma1Resume || mState == State::kSentSigma3;
+    }
+
+    if (mRole == CryptoContext::SessionRole::kResponder)
+    {
+        return mState == State::kSentSigma2 || mState == State::kSentSigma2Resume;
+    }
 
+    return false;
+}
+
+void CASESession::Finish()
+{
     Transport::PeerAddress address = mExchangeCtxt->GetSessionHandle()->AsUnauthenticatedSession()->GetPeerAddress();
 
     // Discard the exchange so that Clear() doesn't try closing it. The exchange will handle that.
```

Comment:
There is no more Clear(); this comment should refer to the destructor now, right?  Same for a bunch of other comments.